core/src/main/scala/com/mesosphere/usi/core/SchedulerLogic.scala

```diff
@@ -0,0 +1,201 @@
+    */
+  private def applyAndUpdateCaches(eventFrameEffects: FrameEffects, dirtyPodIds: Set[PodId]): FrameEffects = {
+    frame = frame.applyStateEffects(eventFrameEffects)
+    this.cachedPendingLaunch = this.cachedPendingLaunch.update(frame, dirtyPodIds)
+
+    val pruneEffects = SchedulerLogic.pruneTaskStatuses(frame)(dirtyPodIds)
+    frame = frame.applyStateEffects(pruneEffects)
+
+    val reviveEffects =
+      if (cachedPendingLaunch.pendingLaunch.nonEmpty)
+        FrameEffects.empty.withMesosCall(Mesos.Call.Revive)
+      else
+        FrameEffects.empty
+
+    eventFrameEffects ++ pruneEffects ++ reviveEffects
+  }
+
+  /**
+    * Process a Mesos event and update internal state.
+    *
+    * @param event
+    * @return The events describing state changes as Mesos call intents
+    */
+  def processMesosEvent(event: Mesos.Event): FrameEffects = {
+    val result = SchedulerLogic.handleMesosEvent(frame, cachedPendingLaunch.pendingLaunch)(event)
+
+    applyAndUpdateCaches(result.frameEffects, result.dirtyPodIds)
+  }
+}
+
+/**
+  * The current home for USI business logic
+  */
+private[core] object SchedulerLogic {
+  private def terminalOrUnreachable(status: PodStatus): Boolean = {
+    // TODO - temporary stub implementation
+    status.taskStatuses.values.forall(status => status == Mesos.TaskStatus.TASK_RUNNING)
+  }
+
+  private def taskIdsFor(pod: PodSpec): Seq[TaskId] = {
+    // TODO - temporary stub implementation
+    Seq(pod.id.value)
+  }
+
+  /**
+    * Process the modification of some podSpec.
+    *
+    * - If a podSpec is deleted:
+    *   - Also delete the podRecord.
+    *   - Prune a terminal / unknown task status.
+    * - If a podSpec is marked as terminal, then issue a kill.
+    *
+    * @return The effects of launching this offer
+    */
+  private def matchOffer(offer: Mesos.Event.Offer, pendingLaunchPodSpecs: Seq[PodSpec]): FrameEffects = {
+    var effects = FrameEffects.empty
+
+    val operations = pendingLaunchPodSpecs.iterator.flatMap { pod =>
+      taskIdsFor(pod).iterator.map { taskId =>
+        Mesos.Operation(Mesos.Launch(Mesos.TaskInfo(taskId)))
+      }
+    }.to[Seq]
+
+    pendingLaunchPodSpecs.foreach { pod =>
+      effects = effects.withPodRecord(pod.id, Some(PodRecord(pod.id, Instant.now(), offer.agentId)))
+    }
+    effects = effects.withMesosCall(Mesos.Call.Accept(offer.offerId, operations = operations))
+
+    effects
+  }
+
+  private[core] def pendingLaunch(goal: Goal, podRecord: Option[PodRecord]): Boolean = {
+    (goal == Goal.Running) && podRecord.isEmpty
+  }
+
+  private[core] def computeNextStateForPods(frame: Frame)(changedPodIds: Set[PodId]): FrameEffects = {
+    changedPodIds.foldLeft(FrameEffects.empty) { (initialEffects, podId) =>
+      var effects = initialEffects
+
+      frame.podSpecs.get(podId) match {
+        case None =>
+          // TODO - this should be spurious if the podStatus is non-terminal
+          if (frame.podStatuses.get(podId).exists { status =>
```

Comment:
I'm not sure I understand what case is handled here - changed `podId` isn't part of the `podSpecs` but has a `podStatus` so we handle it as a terminal or unreachable?

Reply:
It's correct. We clean up terminal or unreachable statuses if the podSpec is removed, and the status is terminal or unreachable.